File: core/src/de/iss/mv2/messaging/DEF_MESSAGE_FIELD.java
package de.iss.mv2.messaging;


/**
 * An enumeration containing all well known message fields.
 * @author Marcel Singer
 *
 */
public enum DEF_MESSAGE_FIELD {
	/**
	 * An unknown message field.
	 */
	UNKNOWN(-1),
	/**
	 * A message field containing the content encoding of a message.
	 */
	CONTENT_ENCODING(1),
	/**
	 * A message field declaring the cause for a message.
	 */
	CAUSE(2),
	/**
	 * A message field containing the symmetric encryption key and IV.
	 */
	ENCRYPTION_KEY(3, CONTENT_TYPE.BINARY),
	/**
	 * A message field containing the name of the symmetric algorithm.
	 */
	SYMMETRIC_ALGORITHM(4),
	/**
	 * A message field containing the name of the asymmetric algorithm.
	 */
	ASYMMETRIC_ALGORITHM(5),
	/**
	 * A message field containing pain text content.
	 */
	CONTENT_PLAIN(6),
	/**
	 * A message field containing the name of a hash algorithm.
	 */
	HASH_ALGORITHM(7),
	/**
	 * A message field containing binary content data.
	 */
	CONTENT_BINARY(8, CONTENT_TYPE.BINARY),
	/**
	 * A message field containing binary hash data.
	 */
	HASH_BINARY(9, CONTENT_TYPE.BINARY),
	/**
	 * A message field containing the receiving web space address.
	 */
	RECEIVER(10),
	/**
	 * A message field containing the subject of a {@link STD_MESSAGE#CONTENT_MESSAGE}.
	 */
	SUBJECT(11),
	/**
	 * A message field containing the senders certificate. The encoded certificate is stored <b>binary</b>.
	 */
	SENDER_CERTIFICATE(12, CONTENT_TYPE.BINARY),
	/**
	 * A message field containing the carbon copy (CC) addresses.
	 */
	CARBON_COPY(13),
	/**
	 * The type of a message field containing a signature.
	 */
	SIGNATURE(14, CONTENT_TYPE.BINARY),
	/**
	 * The type of a message field containing the lower bound of a time range.
	 */
	NOT_BEFORE(15),
	/**
	 * The type of a message field containing a timestamp.
	 */
	TIMESTAMP(16, CONTENT_TYPE.BINARY);
	
	
	/**
	 * Contains the identifier of the message field.
	 */
	private final int identifier;
	
	/**
	 * The type of the field content.
	 */
	private final CONTENT_TYPE contentType;
	
	/**
	 * Creates a new field type.
	 * @param identifier The identifier of the field type to create.
	 */
	private DEF_MESSAGE_FIELD(int identifier) {
		this(identifier, CONTENT_TYPE.STRING_LITERAL);
	}
	
	/**
	 * Creates a new field type.
	 * @param identifier The identifier 
	 * @param contentType
	 */
	private DEF_MESSAGE_FIELD(int identifier, CONTENT_TYPE contentType){
		this.identifier = identifier;
		this.contentType = contentType;
	}
	
	/**
	 * Returns the identifier of this field.
	 * @return The identifier of this field.
	 */
	public int getIdentifier(){ return identifier; }
	
	/**
	 * Returns the content type of the message field.
	 * @return The content type of the message field.
	 */
	public CONTENT_TYPE getContentType(){ return contentType; }
	
	/**
	 * Searches all well known message fields for the given identifier.
	 * @param identifier The identifier to search for.
	 * @return The field type with the given identifier or {@link DEF_MESSAGE_FIELD#UNKNOWN} if no type was found.
	 */
	public static DEF_MESSAGE_FIELD find(int identifier){
		for(DEF_MESSAGE_FIELD f : DEF_MESSAGE_FIELD.values()){
			if(f.getIdentifier() == identifier) return f;
		}
		return UNKNOWN;
	}
	
}

File: WebClient/MV2CorePortable/Messaging/DEF_MESSAGE_FIELD.cs
using System;
using System.Collections.Generic;

using System.Text;

namespace ISS.MV2.Messaging {
    public class DEF_MESSAGE_FIELD {

        private int identifier = -1;
        private bool isBinary;
        private string name;

        public bool IsBinary { get { return isBinary; } }
        public int Identifier { get { return identifier; } }

        public static readonly DEF_MESSAGE_FIELD UNKNOWN = new DEF_MESSAGE_FIELD(-1, "UNKNOWN");
        public static readonly DEF_MESSAGE_FIELD CONTENT_ENCODING = new DEF_MESSAGE_FIELD(1, "CONTENT_ENCODING");
        public static readonly DEF_MESSAGE_FIELD CAUSE = new DEF_MESSAGE_FIELD(2, "CAUSE");
        public static readonly DEF_MESSAGE_FIELD ENCRYPTION_KEY = new DEF_MESSAGE_FIELD(3, true, "ENCRYPTION_KEY");
        public static readonly DEF_MESSAGE_FIELD SYMMETRIC_ALGORITHM = new DEF_MESSAGE_FIELD(4, "SYMMETRIC_ALGORITHM");
        public static readonly DEF_MESSAGE_FIELD ASYMMETRIC_ALGORITHM = new DEF_MESSAGE_FIELD(5, "ASYMMETRIC_ALGORITHM");
        public static readonly DEF_MESSAGE_FIELD CONTENT_PLAIN = new DEF_MESSAGE_FIELD(6, "CONTENT_PLAIN");
        public static readonly DEF_MESSAGE_FIELD HASH_ALGORITHM = new DEF_MESSAGE_FIELD(7, "HASH_ALGORITHM");
        public static readonly DEF_MESSAGE_FIELD CONTENT_BINARY = new DEF_MESSAGE_FIELD(8, true, "CONTENT_BINARY");
        public static readonly DEF_MESSAGE_FIELD HASH_BINARY = new DEF_MESSAGE_FIELD(9, true, "HASH_BINARY");
        public static readonly DEF_MESSAGE_FIELD RECEIVER = new DEF_MESSAGE_FIELD(10, "RECEIVER");
        public static readonly DEF_MESSAGE_FIELD SUBJECT = new DEF_MESSAGE_FIELD(11, "SUBJECT");
        public static readonly DEF_MESSAGE_FIELD SENDER_CERTIFICATE = new DEF_MESSAGE_FIELD(12, true, "SENDER_CERTIFICATE");
        public static readonly DEF_MESSAGE_FIELD CARBON_COPY = new DEF_MESSAGE_FIELD(13, "CARBON_COPY");
        public static readonly DEF_MESSAGE_FIELD SIGNATURE = new DEF_MESSAGE_FIELD(14, true, "SIGNATURE");
        public static readonly DEF_MESSAGE_FIELD NOT_BEFORE = new DEF_MESSAGE_FIELD(15, "NOT_BEFORE");
        public static readonly DEF_MESSAGE_FIELD TIMESTAMP = new DEF_MESSAGE_FIELD(16, true, "TIMESTAMP");

        public static readonly DEF_MESSAGE_FIELD[] WELL_KNOWN_TYPES = new DEF_MESSAGE_FIELD[] {UNKNOWN, CONTENT_ENCODING, CAUSE, ENCRYPTION_KEY, SYMMETRIC_ALGORITHM, ASYMMETRIC_ALGORITHM, CONTENT_PLAIN, HASH_ALGORITHM, CONTENT_BINARY, HASH_BINARY, RECEIVER, SUBJECT, SENDER_CERTIFICATE, 
        CARBON_COPY, SIGNATURE, NOT_BEFORE, TIMESTAMP};


        private DEF_MESSAGE_FIELD(int identifier, bool isBinary, string name) {
            this.identifier = identifier;
            this.isBinary = isBinary;
            this.name = name;
        }

        private DEF_MESSAGE_FIELD(int identifier, string name)
            : this(identifier, false, name) {
        }

        public static DEF_MESSAGE_FIELD Find(int identifier) {
            foreach(DEF_MESSAGE_FIELD mf in WELL_KNOWN_TYPES){
                if(mf.Identifier == identifier) return mf;
            }
            return DEF_MESSAGE_FIELD.UNKNOWN;
        }

        public override string ToString() {
            return name + "[" + identifier + "]";
        }

    }
}
